contracts/protocore/Protocore.sol

```diff
@@ -114,4 +141,102 @@ contract Protocore is MosaicVersion, ValidatorSet, CoconsensusModule {
             openKernelHash
         );
     }
+
+
+    /** Internal Functions */
+
+    /**
+     * @notice proposeLinkInternal() function proposes a valid link to be
+     *         voted later by active validators.
+     *
+     * \pre `parentVoteMessageHash` is not 0.
+     * \pre `parentVoteMessageHash` refers to an already proposed link which
+     *      `targetFinalisation` is at least justified.
+     * \pre `targetBlockHash` is not 0
+     * \pre `targetBlockNumber` is a multiple of the epoch length.
+     * \pre `targetBlockNumber` is bigger than a targetBlockNumber pointed
+     *      by `_parentVoteMessageHash` link.
+     * \pre A vote message hash (calculated with input params) does not exist.
+     *
+     * \post The link is saved in `links` mapping with currently
+     *       open kernel/metablock height as `proposedMetablockHeight`.
+     * \post `targetFinalisation` is set to 'Registered'.
+     * \post forwardVoteCount -s set to 0.
+     */
+    function proposeLinkInternal(
+        bytes32 _parentVoteMessageHash,
+        bytes32 _sourceTransitionHash,
+        bytes32 _targetBlockHash,
+        uint256 _targetBlockNumber
+    )
+        internal
+    {
+        require(
+            _parentVoteMessageHash != bytes32(0),
+            "Parent vote message hash is 0."
+        );
+
+        Link storage parentLink = links[_parentVoteMessageHash];
+
+        require(
+            parentLink.targetBlockHash != bytes32(0),
+            "Parent link does not exist."
+        );
+
+        require(
+            parentLink.targetFinalisation >= CheckpointFinalisationStatus.Justified,
+            "Parent link's target finalisation status should be at least justified."
+        );
+
+        require(
```

Comment:
moving these checks up would save gas on reading from storage; but not so important :D

Reply:
True. Improved!